### PR TITLE
Increasing robustness of adjoint plan optimizations

### PR DESCRIPTION
**Pull request by gaurav-arya**
@devmotion although I was unable to reproduce your concern in https://github.com/JuliaMath/AbstractFFTs.jl/pull/113#discussion_r1315782241, I've tried to make some tweaks to safeguard against any possible type issues in adjoint plans, and also added a test of the promotion behaviour.

Please let me know if you have any remaining concerns -- since this is still blocking release I would also be happy to revert the optimization implemented in #113 if you have concerns about the design that aren't easy to address.

### CODE REVIEW

File: src/definitions.jl [hunk 1]
```diff
@@ -683,10 +683,11 @@ function adjoint_mul(p::Plan{T}, x::AbstractArray, ::FFTAdjointStyle) where {T}
     pinv = inv(p)
     # Optimization: when pinv is a ScaledPlan, check if we can avoid a loop over x.
     # Even if not, ensure that we do only one pass by combining the normalization with the plan.
+    RT = AbstractArray{<:T} # canonicalize eltype of returned array to be extra safe about type stability  
     if pinv isa ScaledPlan && pinv.scale == N
-        return pinv.p * x
+        return convert(RT, pinv.p * x)
     else
-        return (1/N * pinv) * x
+        return convert(RT, (inv(N) * pinv) * x)
```
Comment (devmotion) [inline src/definitions.jl:690]: See my comment here for why I think my initial concern was unjustified: https://github.com/JuliaMath/AbstractFFTs.jl/pull/113/files#r1367787821

However, there could be actual type stability issues in cases where `pinv` is a `ScaledPlan` where `pinv.scale` is not of type `eltype(pinv.p)`. This case is actually intended in the design of `ScaledPlan` and documented in its constructor. It does not matter as long as `typeof(pinv.scale)` does not cause promotions of `eltype(pinv.p)` (such as in the intended case of `Float64` and `Complex{Float64}`), but such a non-promoting behaviour is not guaranteed anywhere in the definition of `ScaledPlan`. Hence I wonder if it's worth it at all to introduce this branch here. Did you perform any benchmarks and see significant speed-ups in practical scenarios?

So maybe we could just simplify this to
```suggestion
  return (inv(N) * pinv) * x
```

File: src/definitions.jl [hunk 2]
```diff
@@ -698,7 +699,7 @@ function adjoint_mul(p::Plan{T}, x::AbstractArray, ::RFFTAdjointStyle) where {T<
     pinv = inv(p)
     n = size(pinv, halfdim)
     # Optimization: when pinv is a ScaledPlan, fuse the scaling into our map to ensure we do not loop over x twice.
-    scale = pinv isa ScaledPlan ? pinv.scale / 2N : 1 / 2N
+    scale = pinv isa ScaledPlan ? pinv.scale / 2N : one(pinv.scale) / 2N
```
Comment (devmotion) [inline src/definitions.jl:702]: This is wrong, the second branch will fail (and is apparently not tested): If `pinv` is not a `ScaledPlan`, `pinv.scale` does not exist in general.
```suggestion
    scale = pinv isa ScaledPlan ? pinv.scale / 2N : inv(2N)
```
Can you add a test that hits the second branch?

File: src/definitions.jl [hunk 3]
```diff
@@ -721,7 +722,7 @@ function adjoint_mul(p::Plan{T}, x::AbstractArray, ::IRFFTAdjointStyle) where {T
     pinv = inv(p)
     d = size(pinv, halfdim)
     # Optimization: when pinv is a ScaledPlan, fuse the scaling into our map to ensure we do not loop over x twice.
-    scale = pinv isa ScaledPlan ? pinv.scale / N : 1 / N
+    scale = pinv isa ScaledPlan ? pinv.scale / N : one(pinv.scale) / N
```
Comment (devmotion) [inline src/definitions.jl:725]:
```suggestion
    scale = pinv isa ScaledPlan ? pinv.scale / N : inv(N)
```

File: src/definitions.jl (hunk 2 above)
Comment (gaurav-arya) [reply src/definitions.jl]: Woah, thank you for the catch.

File: src/definitions.jl (hunk 1 above)
Comment (gaurav-arya) [reply src/definitions.jl]: Agreed: we can keep the one pass optimization but avoid the zero-pass optimization that involves this tricky runtime branching.

**Review by devmotion — APPROVED** [review]

**Pull request merged**